### PR TITLE
NFC: make headers compatible with C++20

### PR DESCRIPTION
**Pull request by redsun82**
In a downstream project we are using the Swift code as a library and compiling with C++20. This mostly works except for two backward incompatibilities introduced by C++20:

* [P1008] disallows aggregate initialization when defaulted or deleted constructors are specified. This lead to a compilation error for aggregate initialization of `swift::Compilation::Result`. The backward and forward compatible fix for that is to provide a constructor turning aggregate initializations into a normal constructor call.
  * **EDIT:** this was fixed recently by https://github.com/apple/swift/pull/67683, so this part of the PR became obsolete, though it was updated to introduce a small optimization of that change using move semantics instead of copy for the pass-by-value `depGraph` parameter (that ultimately holds two `std::unordered_map`s) of `swift::Compilation::Result`'s constructor.
* [P1185] introduces more candidates for overload resolution of comparison operator calls. As explained in [P1630], this leads in some cases to ambiguity compilation errors in C++20, which is exactly the case with `swift::ValueOwnershipKind`. The fix in this case is to remove some redundant operator definitions conditionally on the `__cpp_impl_three_way_comparison` feature test macro, which [includes the papers mentioned above][feature_test].

[P1008]: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1008r1.pdf
[P1185]: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1185r2.html
[P1630]: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1630r1.html
[feature_test]: https://en.cppreference.com/w/cpp/feature_test

### CODE REVIEW

**Comment by redsun82** [comment]
@allevato this PR ties to C++20 compatibility that I've seen you have been working on recently, so I think you might be interested in having a look :slightly_smiling_face:

**Comment by allevato** [comment]
@swift-ci please test

File: include/swift/Driver/Compilation.h [hunk 1]
```diff
@@ -102,8 +102,7 @@ class Compilation {
 
     /// Construct a \c Compilation::Result from just an exit code.
     static Result code(int code) {
-      return Compilation::Result{false, code,
-                                 fine_grained_dependencies::ModuleDepGraph()};
+      return Compilation::Result{false, code};
```
[the remark this answers is not in the document]
Comment (redsun82) [reply include/swift/Driver/Compilation.h:105]: ah, no, this is me not testing the git merge I did :sweat_smile: I had default arguments set in the constructor before

[the remark this answers is not in the document]
Comment (redsun82) [reply include/swift/Driver/Compilation.h]: can you retest now?

**Review by allevato — APPROVED** [review]
Thanks for the fix!

**Pull request merged**